src/test/scala/components/AsyncMemoryUnitTest.scala

```diff
@@ -0,0 +1,194 @@
+// Unit tests for the async memory objects.
+
+package dinocpu
+
+import chisel3.iotesters
+import chisel3.iotesters.{ChiselFlatSpec, Driver, PeekPokeTester}
+
+// To generate random latencies
+import scala.util.Random
+
+
+// Test instruction port reading a zeroed memory file
+class AsyncMemoryUnitTester$IMemZero(m: DualPortedAsyncMemory, size: Int, latency: Int) extends PeekPokeTester(m) {
+  // Expect 0's on the instruction port
+  for (i <- 0 to size/4 - 1) {
+    poke(m.io.imem.address, i*4)
+    poke(m.io.imem.ready, 1)
+
+    step(latency)
+    
+    expect(m.io.imem.valid, 1)
+    expect(m.io.imem.instruction, 0)
+  }
+}
+
+// Test instruction port reading an ascending memory file
+class AsyncMemoryUnitTester$IMemRead(m: DualPortedAsyncMemory, size: Int, latency: Int) extends PeekPokeTester(m) {
+  // Expect ascending bytes on instruction port
+  for (i <- 0 to size/4 - 1) {
+    poke(m.io.imem.address, i*4)
+    poke(m.io.imem.ready, 1)
+    
+    step(latency)
+    
+    expect(m.io.imem.valid, 1)
+    expect(m.io.imem.instruction, i)
+  }
+}
+
+// Test data port writes and instruction port reads
+class AsyncMemoryUnitTester$IMemWrite(m: DualPortedAsyncMemory, size: Int, latency: Int) extends PeekPokeTester(m) {
+  // write ascending data to memory 
+  for (i <- 0 to size/8 - 1) {
+    poke(m.io.dmem.address, i*4)
+    poke(m.io.dmem.memwrite, 1)
+    poke(m.io.dmem.maskmode, 2)
+    poke(m.io.dmem.sext, 0)
+    poke(m.io.dmem.writedata, i+100)
+
+    step(latency)
+
+    expect(m.io.dmem.valid, 1)
+  }
+
+  poke (m.io.dmem.memwrite, 0)
+
+  // expect ascending bytes on instruction port
+  for (i <- 0 to size/4 - 1) {
```

Comment:
Instead of having `-1` you can use `until` instead of `to`, I believe.

Reply:
I can make that change really quick.